libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,200 @@
+
+	return std::move(builder.m_graph);
+}
+
+bool FunctionCallGraphBuilder::visit(Identifier const& _identifier)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	{
+		solAssert(*_identifier.annotation().requiredLookup == VirtualLookup::Virtual, "");
+
+		auto funType = dynamic_cast<FunctionType const*>(_identifier.annotation().type);
+
+		// For events kind() == Event, so we have an extra check here
+		if (funType && funType->kind() == FunctionType::Kind::Internal)
+		{
+			processFunction(callable->resolveVirtual(*m_contract), _identifier.annotation().calledDirectly);
+
+			solAssert(m_currentNode.has_value(), "");
+		}
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(NewExpression const& _newExpression)
+{
+	if (ContractType const* contractType = dynamic_cast<ContractType const*>(_newExpression.typeName().annotation().type))
+		m_graph->createdContracts.emplace(&contractType->contractDefinition());
+
+	return true;
+}
+
+void FunctionCallGraphBuilder::endVisit(MemberAccess const& _memberAccess)
+{
+	auto functionType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type);
+	auto functionDef = dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration);
+	if (!functionType || !functionDef || functionType->kind() != FunctionType::Kind::Internal)
+		return;
+
+	// Super functions
+	if (*_memberAccess.annotation().requiredLookup == VirtualLookup::Super)
+	{
+		if (auto const* typeType = dynamic_cast<TypeType const*>(_memberAccess.expression().annotation().type))
+			if (auto const contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+			{
+				solAssert(contractType->isSuper(), "");
+				functionDef =
+					&functionDef->resolveVirtual(
+						*m_contract,
+						contractType->contractDefinition().superContract(*m_contract)
+					);
+			}
+	}
+	else
+		solAssert(*_memberAccess.annotation().requiredLookup == VirtualLookup::Static, "");
+
+	processFunction(*functionDef, _memberAccess.annotation().calledDirectly);
+	return;
+}
+
+void FunctionCallGraphBuilder::endVisit(ModifierInvocation const& _modifierInvocation)
+{
+	if (auto const* modifier = dynamic_cast<ModifierDefinition const*>(_modifierInvocation.name().annotation().referencedDeclaration))
+	{
+		if (*_modifierInvocation.name().annotation().requiredLookup == VirtualLookup::Virtual)
+			modifier = &modifier->resolveVirtual(*m_contract);
+		else
+			solAssert(*_modifierInvocation.name().annotation().requiredLookup == VirtualLookup::Static, "");
+
+		processFunction(*modifier);
+	}
+}
+
+void FunctionCallGraphBuilder::visitCallable(CallableDeclaration const* _callable, bool _directCall)
+{
+	solAssert(!m_graph->edges.count(_callable), "");
+
+	std::optional<Node> previousNode = m_currentNode;
+	m_currentNode = _callable;
+
+	if (previousNode.has_value() && _directCall)
+		add(*previousNode, _callable);
+	if (!_directCall)
```

Comment:
``_directCall`` can only be ``false`` when this is called from ``processFunction``, right? If so we could move this part to ``processFunction`` and remove the ``_directCall`` argument from this function.

Reply:
No, the if check in `processFunction` only applies to the `add()` call. The `visitCallable()` call will always be done and can have either true or false as parameter value

Reply:
I think we're speaking past each other again :-). I mean always assume that ``visitCallable`` calls directly and instead adjust the previous node to the dispatch in ``processFunction`` (making it a "direct call from the dispatch").

I.e. wouldn't the following work:
```
void FunctionCallGraphBuilder::visitCallable(CallableDeclaration const* _callable)
{
 	solAssert(!m_graph->edges.count(_callable), "");
	if (m_currentNode.has_value())
		add(*m_currentNode, _callable);

	m_currentNode = _callable;
    _callable->accept(*this); 
}

void FunctionCallGraphBuilder::processFunction(CallableDeclaration const& _callable, bool _calledDirectly)
{
	if (m_graph->edges.count(&_callable))
		return;

    std::optional<Node> previousNode = m_currentNode;
	// Create edge to creation dispatch
	if (!_calledDirectly)
    {
        add(*m_currentNode, m_currentDispatch);
        m_currentNode = m_currentDispatch;
     }

	visitCallable(&_callable);
    m_currentNode = previousNode;
}
```

as well as setting ``m_currentNode = SpecialNode::Entry;`` in the constructor instead of manually adding the edge in the ``interfaceFunctionList`` loop?

Not that I needlessly want to have you move around stuff - but currently especially in the absence of any comments it looks pretty random which function connects which node to which given what flag... my suggestion would clearly define:
``visitCallable`` - appends the current callable to the current call graph node, if any.
``processFunction`` - adds a (direct or indirect) function call from the current function to a given function (and does not affect the current node)

But yeah - also no problem to leave it as is, I just find it a bit convoluted myself, but it may just be a matter of taste :-).

Reply:
I have actually made the this exact change even before I found this comment :) Now it works very similarly to what @ekpyron suggested.

The main difference is that I have moved the `add()` call to `processFunction()` too because otherwise we'd be skipping any edges **to** the callable once it gets at least one edge coming **from** it. By the way, @chriseth, this is actually a good example of a sneaky bug that would be pretty hard to find if we only checked whether the graph does contain all functions that the codegen includes. With my test suite I discovered it as one of the first things.

Reply:
I see that the bug was actually pointed out by @ekpyron in the [next comment](https://github.com/ethereum/solidity/pull/10332#discussion_r555042512) ;)